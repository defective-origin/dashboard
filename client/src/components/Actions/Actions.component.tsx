import React from 'react'
// TODO: (tikhon_puntus) remove all react imports

// ---| core |---
import { cn } from 'tools'

// ---| pages |---
// ---| screens |---
// ---| components |---
import Link from 'components/Link'
import Button from 'components/Button'
import Repeat, { ComponentWithItems, RepeatItem } from 'components/Repeat'
import Block, { BlockProps } from 'components/Block'

// ---| self |---
import css from './Actions.module.scss'

export const ACTION_MAP = {
  button: Button,
  link: Link,
}

export type ActionItem = RepeatItem<typeof ACTION_MAP>

export type ActionsProps = ComponentWithItems<BlockProps, ActionItem>

/**
 * Component description.
 *
 * How to use
 * @example
 * <Actions />
 */
export function Actions(props: ActionsProps): JSX.Element {
  const { items, children, className, ...otherProps } = props
  const _className = cn(css.Actions, className)

  return (
    <Block className={_className} align='center' direction='x' {...otherProps}>
      <Repeat cmp={ACTION_MAP} items={items} variant='button' />

      {children}
    </Block>
  )
}

Actions.displayName = 'Actions'

Actions.Link = Link
Actions.Button = Button

export default Actions
