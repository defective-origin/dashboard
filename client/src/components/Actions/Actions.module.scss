@import 'src/App/App.module';

:global {
  .dark {
    // --text-color: white;
  }

  .light {
    // --text-color: black;
  }
}

.Actions {
  padding: 5px;
  display: flex;
}

.x {
  flex-direction: row;
  align-items: center;
}

.y {
  flex-direction: column;
  justify-content: center;
}

.xy {
  flex-wrap: wrap;
}
