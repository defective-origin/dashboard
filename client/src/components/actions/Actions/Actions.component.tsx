import React from 'react'

// ---| core |---
import { cn } from 'tools'
import { Size } from 'theme'

// ---| pages |---
// ---| screens |---
// ---| components |---
import Menu, { MenuProps } from 'components/actions/Menu'
import Link from 'components/actions/Link'
import Button from 'components/actions/Button'
import Divider from 'components/layouts/Divider'
import NavLink from 'components/actions/NavLink'
import { PropsWithItems, RepeatItem } from 'components/layouts/Repeat'
import Block, { BlockProps } from 'components/layouts/Block'

// ---| self |---
import css from './Actions.module.scss'

export const ACTION_MAP = {
  divider: Divider,
  button: Button,
  link: Link,
  nav: NavLink,
  menu: Menu,
}

export type GenericActionItem = RepeatItem<typeof ACTION_MAP>
export type CustomActionItem = React.ElementType | React.ReactNode
export type ActionWithMenu<T> = (T & { items?: ActionWithMenu<T>, hide?: boolean })
export type ActionItem = ActionWithMenu<GenericActionItem> | CustomActionItem

export type ActionsProps = PropsWithItems<ActionItem, BlockProps> & {
  menu?: MenuProps['v']
  size?: Size
}

/**
 * Component description.
 *
 * How to use
 * @example
 * const items: ActionItem[] = [
 *   <CustomItem />,
 *   <Divider v='x />,
 *   <Button v='outlined' content='ITEM 1' color='error', hide />,
 *   { variant: 'nav', v: 'outlined', content: 'ITEM 1', color: 'error', hide: true },
 *   { variant: 'link', v: 'outlined', content: 'ITEM 2', color: 'warning' },
 *   { variant: 'button', v: 'outlined', content: 'ITEM 3', color: 'info', items: [
 *      { variant: 'button', v: 'outlined', content: 'SUB ITEM 1', color: 'error', hide: true },
 *      { variant: 'button', v: 'outlined', content: 'SUB ITEM 3', color: 'warning' },
 *   ]},
 * ]
 *
 * <Actions items={items} menu='left' />
 */
export function Actions(props: ActionsProps) {
  const { size, menu, items, children, className, ...otherProps } = props
  const _className = cn(css.Actions, className)
  const menuItems = items?.map((item, idx) => {
    // custom element
    if (React.isValidElement(item) || !item) {
      return <React.Fragment key={idx}>{item}</React.Fragment>
    }

    const { variant = 'button', hide, tooltip, items, ...otherItemProps} = item as any
    const Tag = ACTION_MAP[variant as 'button']

    // hide by condition
    if (hide) {
      return null

    // submenu
    } else if (items) {
      return (
        <Menu
          key={idx}
          v={menu ?? 'bottom'}
          items={items}
          trigger={o => (
            <Tag
              size={size}
              active={o.isOn}
              {...otherItemProps}
            />
          )}
        />
      )
    }

    return <Tag key={idx} tooltip={tooltip} tooltipSide={menu} size={size} {...otherItemProps} />
  })

  return (
    <Block className={_className} aligns='center' v='x' {...otherProps}>
      {menuItems}
      {children}
    </Block>
  )
}

Actions.displayName = 'Actions'

Actions.Link = Link
Actions.Nav = NavLink
Actions.Button = Button
Actions.Menu = Menu

export default Actions
