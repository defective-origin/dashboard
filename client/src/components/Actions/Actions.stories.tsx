/* eslint-disable no-restricted-imports */
import { params, field } from '../../../.storybook/tool'
import type { Meta, StoryObj } from '@storybook/react'
import Actions, { ActionItem } from './Actions.component'

const ITEMS: ActionItem[] = [
  { variant: 'button', start: 'settings', end: 'settings', tooltip: 'Television', content: 'button', v: 'outlined' },
  { variant: 'link', start: 'settings', tooltip: 'Add Widget', content: 'link' },
  {
    start: 'settings', end: 'settings', content: 'button menu', v: 'outlined', items: [
      { start: 'settings', content: 'aaa', tooltip: 'aaa' },
      { start: 'settings', content: 'bbb', tooltip: 'bbb' },
      {
        start: 'settings', content: 'ccc', items: [
          { start: 'settings', content: 'ddd', tooltip: 'ddd' },
          { start: 'settings', content: 'eee', tooltip: 'eee' },
          {
            start: 'settings', content: 'fff', items: [
              { start: 'settings', content: 'ggg', tooltip: 'ggg' },
              { start: 'settings', content: 'hhh', tooltip: 'hhh' },
              { start: 'settings', content: 'yyy', tooltip: 'yyy' },
            ],
          },
        ],
      },
    ],
  },
]

const meta: Meta<typeof Actions> = {
  component: Actions,
  title: 'Components/ACTIONS/Actions',
  tags: ['autodocs'],
  argTypes: {
    menu: field.variants(['top', 'right', 'bottom', 'left'], 'MenuVariant'),
  },
}

export default meta

type Story = StoryObj<typeof Actions>

export const Demo: Story = {
  parameters: params('Actions'),
  args: {
    items: ITEMS,
    menu: 'top',
    g: 'xxl',
  },
  render: (props) => <Actions items={ITEMS} {...props} />,
}
