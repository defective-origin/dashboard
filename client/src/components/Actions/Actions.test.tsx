// ---| tests |---
import { render } from 'tests'

// ---| self |---
import Actions from './Actions.component'

describe('[Actions] component', () => {
  it('should render component', () => {
    const container = render(<Actions />)

    expect(container.snapshot()).toMatchSnapshot()
  })
})
