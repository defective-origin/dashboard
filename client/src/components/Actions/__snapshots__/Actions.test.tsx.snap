// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`[Actions] component > should render component 1`] = `
<DocumentFragment>
  <div
    class="block block--x item _Actions_dd10e3"
    style="align-items: center;"
  />
</DocumentFragment>
`;
